src/main/kotlin/sink/KafkaSink.kt

```diff
@@ -1,46 +1,79 @@
+import org.apache.kafka.common.header.internals.RecordHeader
+import java.util.*
 
+private val log = KotlinLogging.logger("kafka-sink")
 
-private fun mkProducer(servers: String?) =
-    ProducerBuilder.ofByteArray
-        .option("client.id", "leia")
-        .let {
-            if (servers != null) {
-                it.option("bootstrap.servers", servers)
-            } else it
-        }.create()
+val commonOptions = mapOf("client.id" to "leia")
+
+val consumerOptions = mapOf(
+    "key.deserializer" to "org.apache.kafka.common.serialization.ByteArrayDeserializer",
+    "value.deserializer" to "org.apache.kafka.common.serialization.ByteArrayDeserializer",
+    "enable.auto.commit" to "false")
+
+val producerOptions = mapOf(
+    "key.serializer" to "org.apache.kafka.common.serialization.StringSerializer",
+    "value.serializer" to "org.apache.kafka.common.serialization.ByteArraySerializer",
+    "request.timeout.ms" to "10000",
+    "max.block.ms" to "5000",
+    "acks" to "all",
+    "compression.type" to "gzip",
+    "retries" to "0")
+
+private fun kafkaOptions(servers: String?, producer: Boolean) = Properties()
+    .apply { putAll(commonOptions) }
+    .apply { putAll(if (producer) producerOptions else consumerOptions) }
+    .apply { servers?.let { put("bootstrap.servers", it) } }
+
+private fun mkProducer(servers: String?) = KafkaProducer<String, ByteArray>(kafkaOptions(servers, producer = true))
+
+private fun mkChecker(servers: String?) =
+    KafkaConsumer<String, ByteArray>(kafkaOptions(servers, producer = false))
 
 private fun <K, V> Map<K, V>.with(addend: Pair<K, V?>?): Map<K, V> =
     if (addend?.second != null)
         plus(addend.first to addend.second!!).toMap()
     else this
 
 private class KafkaSink(
-    private val producer: Producer<String, ByteArray>,
-    private val description: SinkDescription
+    private val producer: KafkaProducer<String, ByteArray>,
+    private val description: SinkDescription,
+    private val host: String?
 ) : Sink {
     override suspend fun handle(incomingRequest: IncomingRequest): SinkResult {
         val body = description.dataFormat.generateBody(incomingRequest)
-        val key = incomingRequest.requestId
-        val headers: Map<String, ByteArray> = emptyMap<String, ByteArray>()
+        val key = incomingRequest.requestId.toString()
+        val headers = emptyMap<String, ByteArray>()
             .with("leia/user" to description.authorizedAs?.toByteArray())
+            .map { RecordHeader(it.key, it.value) }
 
         return try {
-            producer.send(description.topic, key.toString(), body, headers)
+            producer.send(ProducerRecord<String, ByteArray>(description.topic, null, key, body, headers))
             SinkResult.SuccessfullyWritten
         } catch (exc: Exception) {
             SinkResult.WritingFailed(exc)
         }
     }
+
+    override suspend fun healthCheck() =
+        try {
+            mkChecker(host)
```

Comment:
Ah does this check for bootstrap servers being up?

Reply:
At lest it would report error if kafka container goes down.

Reply:
👍